Filters out private taxonomies from sidebar UI.

BE-Webdesign opened the pull request:
Fixes #2450,

Could potentially break if the user does not have privileges to see the
edit context. Will need to test more.

**Testing Instructions**

1. Register a private taxonomy.
2. Load Gutenberg.
3. Verify that the taxonomy does not show up in the Categories & Tags portion of the sidebar.

BE-Webdesign committed:
Filters out private taxonomies from sidebar UI.

Fixes #2450,

Adds a public property to the taxonomy field, this is used to determine
whether the taxonomy is public or private. Private taxonomies will now
properly be hidden. Does not work 100% properly for Authors, this is a larger problem with
REST API.

aduth commented on lib/rest-api.php:
```diff
@@ -0,0 +1,44 @@
+<?php
+/**
+ * Internationalization-related functions for the Gutenberg editor plugin.
```
Inaccurate description. Probably something about these being enhancements to existing endpoints? A bit unclear where the line should be drawn between `compat.php` and this one.

BE-Webdesign replied:
Lol woops, copy pasta.

BE-Webdesign replied:
This should probably go into `compat.php`. Didn't realize what that was until now.

aduth commented on editor/sidebar/post-taxonomies/index.js:
```diff
@@ -35,9 +35,11 @@ class PostTaxonomies extends Component {
 
 	componentDidMount() {
 		this.fetchTaxonomies = new wp.api.collections.Taxonomies()
-			.fetch()
```
Maybe not here, but we should see about upgrading this component to using the `withAPIData` higher-order component.

https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-api-data

youknowriad replied:
already done and merged, needs rebase :)

BE-Webdesign replied:
I will rebase and use the component.

aduth commented on editor/sidebar/post-taxonomies/index.js:
```diff
@@ -35,9 +35,11 @@ class PostTaxonomies extends Component {
 			.done( ( taxonomies ) => {
-				this.setState( { taxonomies: Object.values( taxonomies ) } );
+				this.setState(
+					{ taxonomies: Object.values( taxonomies ).filter( taxonomy => taxonomy.public ) }
```
Minor: Lodash's `filter` handles this beautifully (coercing object to arrange, filtering on key):

```js
this.setState( {
	taxonomies: filter( taxonomies, { public: true } ),
} );
```

BE-Webdesign replied:
Was thinking about using Lodash, this looks perfect!

aduth reviewed (commented):
Works well in my testing.

BE-Webdesign commented:
@aduth, @youknowriad. This should be ready now. Rebased to use the new method Riad implemented, using higher order component. Does not fully work for authors, but that is tied to a larger problem.

BE-Webdesign commented:
Fixes #2450,

Adds a public property to the taxonomy field, this is used to determine
whether the taxonomy is public or private. Private taxonomies will now
properly be hidden. Does not work 100% properly for Authors, this is a larger problem with
REST API.

BE-Webdesign commented:
Anybody know why the test failed?

>Fatal error: Class 'WP_Block_Type' not found in /home/travis/build/WordPress/gutenberg/phpunit/class-wp-dummy-block-type.php on line 11

I don't think that has anything to do with this PR. Maybe re start the build?

youknowriad commented:
@BE-Webdesign you should try to rebase, this probably means the build will fail on master if you merge, due to some other PR.

BE-Webdesign commented:
Pretty sure I rebased against master. I will rebase again.

BE-Webdesign commented:
That seemed to fix it, but I dunno why, it didn't bring in any new files etc.

youknowriad commented:
computers 💻 are weird :)

BE-Webdesign commented:
> computers 💻 are weird :)

Yup, this happens all too often 😅 .

youknowriad reviewed (approved):
Works great! I wonder if this fields should not be added by default to the Core Endpoint.

> Does not work 100% properly for Authors, this is a larger problem with REST API.

Could you detail the issue here and maybe create an issue/ticket for it?

Since we decided to start the API work here in Gutenberg and potentially move them to Core later, this looks good to me.

BE-Webdesign commented:
> Could you detail the issue here and maybe create an issue/ticket for it?

#2545 is the ticket explaining most of it. Basically authors cannot view taxonomy data ever, due to permissions.

This PR, should magically work for authors, once #2545 is resolved, but they are somewhat separate issues.

BE-Webdesign commented:
> I wonder if this fields should not be added by default to the Core Endpoint.

It is on the edit context, which is a restricted context, so this will only be visible to authenticated users. I am not sure what other conditional we would want to add to hide it. I think it should be fine as is. But I will wait for rmccue review, before merging.

BE-Webdesign commented:
I think this problem is actually slightly more sinister than previously though. I have to verify but, I think private taxonomies will still display on a post page as long as the have a valid metabox callback set. So maybe we need to use that instead.

pento commented:
I've created a [core ticket](https://core.trac.wordpress.org/ticket/42707) to expose the visibility settings for taxonomies.

Note: To mirror [Core behaviour](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/edit-form-advanced.php?marks=260#L257), you need to test the `show_ui` setting, not the `public` setting.

schlessera committed:
Add visibility property array

This is a polyfill for Core #42707 so that Gutenberg can be tested on older versions of Core.

See https://core.trac.wordpress.org/ticket/42707

schlessera committed:
Adapt taxonomy controller unit tests

schlessera committed:
Filter taxonomies based on visibility.show_ui

schlessera committed:
Add text domain to translations

schlessera committed:
Fix PHPCS alignment issues

schlessera committed:
Adapt test to provide show_ui property

schlessera committed:
Add test to assert that show_ui: false hides taxonomies

schlessera commented:
I did the following changes to update this PR:

- rebased to latest master
- added polyfill to backport https://core.trac.wordpress.org/ticket/42707
- adapted unit tests to check for visibility properties
- adapted post taxonomy component to hide on `visibility.show_ui: false`
- added test to ensure this works as expected

danielbachhuber commented:
I rebased this against master and tested locally with the following:

```
add_action( 'init', function() {
    register_taxonomy( 'secrets', array( 'post' ), array(
        'label'             => 'Secrets',
        'public'            => true,
        'show_ui'           => false,
        'show_in_rest'      => true,
        'rest_base'         => 'secrets',
    ) );
});
```

Using this pull request, the "Secrets" tag UI isn't displayed as expected.

mtias commented:
Nice to see this in!
